evalml/pipelines/components/transformers/column_selectors.py

```diff
@@ -122,3 +123,44 @@ def transform(self, X, y=None):
             pd.DataFrame: Transformed X.
         """
         return super().transform(X, y)
+
+
+class SelectByTypeorTag(ColumnSelector):
+    """
+    Selects columns by specified Woodwork logical type or semantic tag in input data.
+
+    Arguments:
+        columns (list(string), list(ww.LogicalType)): List of datatypes, used to determine which columns to select.
+        random_seed (int): Seed for the random number generator. Defaults to 0.
+    """
+
+    name = "Select Columns By Type or Tag Transformer"
+    hyperparameter_ranges = {}
+    """{}"""
+    needs_fitting = False
+
+    def _check_input_for_columns(self, X):
+        """
+        This check is not required, since ww.select does not throw an error if the requested types are not in the dataset.
+        """
+        return
+
+    def _modify_columns(self, cols, X, y=None):
+        columns = X.ww.select(cols)
+        if len(cols) > 0 and columns.empty:
+            warnings.warn(
+                "No columns of the selected type(s) were found in the input data. SelectByTypeorTag will return an empty DataFrame"
+            )
```

Comment:
Do we want this behavior?  I would think that we might want to take the same route as the sibling classes and raise an exception if a bad selection is made.  @angela97lin what use cases are we trying to satisfy with this?

Reply:
Good thinking @eccabay and great point @chukarsten .

I think we should raise an exception here.

Example: graph filters down to categoricals only, then passes those to OHE and to estimator. If there are no categoricals in the dataset, OHE will be passed empty DF (and possibly error out), estimator will be passed empty DF (and possibly error out) and if for some reason there weren't errors in training the model will not have any value because it didn't receive any input features! This feels to me like grounds for a clear error message from this component.

Reply:
I like the idea of raising an exception or doing nothing at all:

Doing nothing at all: for the component independently, it can arguably be okay if the user wants to filter for a certain type and there happens to be no columns with that type. This is trickier with the component graph, though, as @dsherry pointed out, it is unclear what our other components do with an empty df. If our other components are able to handle empty dfs, I feel like it could be okay to do nothing: in the example above, we could try to have a separate branch for selecting categoricals for the OHE. If there are none, then the OHE doesn't do anything to the empty DF. That can be concatted with the non-categoricals and passed to the estimator. 

TLDR: safety net can be just to raise error but I feel like there's some validity in not doing anything at all /shrug

Reply:
Does anyone know for certain that a downstream component would error out if passed an empty DataFrame? I ask since the default parameter for `SelectColumns` selects no columns and returns an empty DF (and we test to ensure this). Obviously, the situation is a little different if the user attempted to select a certain type and that failed, but I'm unclear how worried we should be about just passing empty DFs through pipelines.

Reply:
Yea it would error out. Our components are not written to handle empty features so I think it's ok to raise an exception! My reasoning for raising a warning instead was that ww lets you select for types/tags that are not present and I figured that our component should not get in the way of that. I don't think we need to worry about passing empty features around. The pipelines created by AutoMLSearch don't do that so that would only happen if the user is doing something custom - we typically let users encounter failures in those cases. Making our components resilient to empty features could be an interesting project, but not one that we need to embark on now?

![image](https://user-images.githubusercontent.com/41651716/127027084-dbd99aed-1b1c-418c-91d3-81f98545afe3.png)

Reply:
Post office hours discussion, moving forward by throwing an error if the resulting DataFrame is empty. I've filed an issue to track further discussion/work in #2554